### PR TITLE
Test non-packed inputs with naive reference convolution kernels

### PR DESCRIPTION
**Pull request by amberhassaan**
This PR follows up #2334 where I fix the testing to code to allow testing non-packed tensors for convolution reference kernel. 
Changes included:
* Modify tensor generation for reference kernel test to expand the strides by multiplying strides with an integer.
* Changes in call stack to allow non-strided inputs to be accepted (some checks disallow non-strided inputs to begin with).

### CODE REVIEW

File: driver/random.hpp
Comment (CAHEK7) [inline driver/random.hpp]:
```suggestion
    assert(B > A);
```
I guess B must not be equal to A, otherwise we may get integer division by 0.

But interesting fact - while B < A does not make much sense, it will work fine for signed integers, since `X%(Y)==X%(-Y) `, but won't work for unsigned integers (at all) and floats (will behave strangely).

**Review by CAHEK7 — CHANGES_REQUESTED** [review]

File: driver/random.hpp
Comment (amberhassaan) [reply driver/random.hpp]: good point. I was thinking about that myself. Initially I though B was inclusive to the range, but now it's clear that it's exclusive.

File: test/gpu_reference_kernel.cpp
Comment (CAHEK7) [inline test/gpu_reference_kernel.cpp]: Is there a plan to get non-strided tensors?

**Review by CAHEK7 — DISMISSED** [review]

File: test/gpu_reference_kernel.cpp
Comment (amberhassaan) [reply test/gpu_reference_kernel.cpp]: These are some initial tests. I have another variant, where I multiply a prefix of strides with a random constant instead of all strides. Need to discuss this with @JehandadKhan and others.

File: test/gpu_reference_kernel.cpp
Comment (junliume) [inline test/gpu_reference_kernel.cpp:316]: @amberhassaan HIP tidy issue:
```
[2023-09-18T16:25:31.036Z] /var/jenkins/workspace/Open_amber_tests-non-packed-conv/test/gpu_reference_kernel.cpp:333:11: warning: Consider using std::transform algorithm instead of a raw loop. [useStlAlgorithm]
[2023-09-18T16:25:31.036Z]         v = v * c;
[2023-09-18T16:25:31.036Z]           ^
```

Comment (CAHEK7) [reply test/gpu_reference_kernel.cpp:316]: Thats exactly I was thinking about but decided to keep it!
Didn't know that clang-tidy does that kind of analysis.

**Review by iq136boy — COMMENTED** [review]
@amberhassaan did you test the reference kernel when g>1?

**Comment by amberhassaan** [comment]
> @amberhassaan did you test the reference kernel when g>1?

Hi @iq136boy , yes, I believe there are tests that provide  a value of `g` that is `> 1`.

File: test/gpu_reference_kernel.cpp
Comment (amberhassaan) [reply test/gpu_reference_kernel.cpp:316]: Tbh, I take issue with this suggestion from `hip-tidy`. The original code is more readable and `std::transform` will obscure it. I would make such a change for a large vector input but that will never happen here. So, I'll silence `hip-tidy` with a comment.

Comment (junliume) [reply test/gpu_reference_kernel.cpp:316]: @amberhassaan okay but please suppress it only for this snippet of code instead of globally. :)

Comment (amberhassaan) [reply test/gpu_reference_kernel.cpp:316]: Yes, that was always my intention. This is a small silly loop that might itself change by the end of this PR.

**Comment by junliume** [comment]
@amberhassaan could you check the "target" branch of this PR?

**Comment by amberhassaan** [comment]
> @amberhassaan could you check the "target" branch of this PR?

Hi @junliume , This is based off of PR #2334 and I have set it as the merge target because I wanted to show the actual diff over 2334 to the reviewers. Once #2334 merges, I'll switch the target to `develop` (or you can do that too).

**Comment by atamazov** [comment]
@amberhassaan 
> Hi @junliume , This is based off of PR #2334 and I have set it as the merge target because I wanted to show the actual diff over 2334 to the reviewers. Once #2334 merges, I'll switch the target to `develop` (or you can do that too).

👍 I think this is proper approach. But please mark this as a draft to avoid accidental merging into the wrong branch.

File: src/ocl/convolutionocl.cpp
Comment (atamazov) [inline src/ocl/convolutionocl.cpp]: Why this hack is necessary? Of course I know that [F8 support needs some improvement](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2251#issuecomment-1728457592).

File: src/solver/conv_direct_naive_conv_bwd.cpp
Comment (atamazov) [inline src/solver/conv_direct_naive_conv_bwd.cpp:137]: [Notice] This is good, but I recommend avoiding making any changes that beyond main PR topic in the future.

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [inline test/gpu_reference_kernel.cpp]: Please remove this or explain in comments why this is necessary to keep.

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [inline test/gpu_reference_kernel.cpp]: Usual advice about doxygen formatting and github username. Also I do not recommend copying comments. But you can use the `\anchor` tag like
```C++
/// \anchor test_naive_gpu_kernel_why_copy_output_before_computation
/// \todo ...
/// ...
```
and then simply
```C++
/// \ref test_naive_gpu_kernel_why_copy_output_before_computation
```

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [inline test/gpu_reference_kernel.cpp:439]: hipMemcpy should not be used with OpenCL backend.

BTW We no longer support OCL backend so you can simply keep this as is or remove the whole OpenCL block of code (I would prefer the former due to [this](https://github.com/ROCmSoftwarePlatform/MIOpen/discussions/1687#discussioncomment-7124363))

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [inline test/gpu_reference_kernel.cpp]: same as https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2394#discussion_r1337264777

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [inline test/gpu_reference_kernel.cpp]: [Q] Can you please explain a bit more why this TODO is necessary? Are you sure that the word "indices" is adequate in this context?

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [inline test/gpu_reference_kernel.cpp]: same as https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2394#discussion_r1337270907

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [inline test/gpu_reference_kernel.cpp]: same as https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2394#discussion_r1337270907

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [inline test/gpu_reference_kernel.cpp]: same as https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2394#discussion_r1337270907

File: src/ocl/convolutionocl.cpp
Comment (amberhassaan) [reply src/ocl/convolutionocl.cpp]: This is necessary to allow accepting non-strided tensors. In fact, we should remove this check all together.

File: src/solver/conv_direct_naive_conv_bwd.cpp
Comment (amberhassaan) [reply src/solver/conv_direct_naive_conv_bwd.cpp:137]: Agree. I just couldn't resist making this change :).

File: test/gpu_reference_kernel.cpp
Comment (amberhassaan) [reply test/gpu_reference_kernel.cpp:439]: Sorry, the hyperlink to `this` above is broken. I don't think I have any `hipMemCpy` outside of `#elif MIOPEN_BACKEND_HIP`, so OpenCL backend is unaffected. I could perhaps add the corresponding `clEnqueWriteBuffer` to OpenCL backend if we care to maintain it. If not, I prefer deleting all `OpenCL` code from this file.

File: test/gpu_reference_kernel.cpp
Comment (amberhassaan) [reply test/gpu_reference_kernel.cpp]: So output is non-packed, i.e. larger than what's needed for convolution and may have data that is not touched by convolution operation. In order to verify the output, GPU buffer should have the same data in the `untouched` parts as the CPU, which is achieved by copying the output to the GPU before the convolution is invoked.

Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: @amberhassaan Have we made sure that untouched areas (gaps) in the reference buffer are zeroed? The random data may contain INF or NAN, etc., and this will spoil the comparison results.

Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: @amberhassaan 
> So output is non-packed, i.e. larger than what's needed for convolution and may have data that is not touched by convolution operation. In order to verify the output, GPU buffer should have the same data in the untouched parts as the CPU, which is achieved by copying the output to the GPU before the convolution is invoked.

[Recommendation] Put this explanation into C++ comments as it seems much more clear to me ;)

In fact I would prefer zeroing both reference and output buffers prior computations, and _only if_ the output tensor is non-packed. There is some MIOpen kernel that can do this job quickly. Is it possible to do?

File: src/ocl/convolutionocl.cpp
Comment (atamazov) [reply src/ocl/convolutionocl.cpp]: [resolved]

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: [resolved]

File: src/solver/conv_direct_naive_conv_bwd.cpp
Comment (atamazov) [reply src/solver/conv_direct_naive_conv_bwd.cpp:137]: [resolved]

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: [resolved]

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp:439]: @amberhassaan URL fixed, sorry.

Comment (atamazov) [reply test/gpu_reference_kernel.cpp:439]: @amberhassaan I prefer keeping this as is, but the final decision is up to you.

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: @amberhassaan Alternatively, we can implement a tensor-layout-aware verification. This is more robust solution, but implementation seems to be much more complicated.

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: [resolved]

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: [resolved]

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: [resolved]

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: [resolved]

@amberhassaan Please hide all [resolved] comments (I can't do that), thanks.

File: test/gpu_reference_kernel.cpp
Comment (CAHEK7) [reply test/gpu_reference_kernel.cpp]: Tensor-layout-aware may help to compare only changed parts, at least as I understood it. But we also want to check that unchanged parts are actually unchanged, and it means that zero-filled tensors are not so good and fp-comparison is not a good idea too. Ideally it must compare the untouched parts as a binary data and doesn't care about all of those `INF`s and `NAN`s.
N.B. as far as I know current random functions do not generate `INF`s and `NAN`s, but there are some direct `NAN` initialiations.

Comment (amberhassaan) [reply test/gpu_reference_kernel.cpp]: @atamazov , @CAHEK7, good suggestions. I find layout-aware verification a bit much given the deadlines for this work. I see two ways forward (and I prefer 1)
1. Fill the output tensor with random integers (that's how we generate inputs). Our verification (against cpu code) compares all data elements (including those not touched by convolution). 
2. Zero-out the entire tensor, which is not as robust as @CAHEK7 pointed out because we're relying on zero being a special value.

File: test/gpu_reference_kernel.cpp
Comment (amberhassaan) [reply test/gpu_reference_kernel.cpp:439]: Discussed this with Jehandad and we decided that it's OK to remove OpenCL code. I can do that in this PR @atamazov if you're OK with that or leave it as a follow up action.

Comment (atamazov) [reply test/gpu_reference_kernel.cpp:439]: @amberhassaan Removing looks good -- as I said, the final decision is up to you.

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: @CAHEK7 
> But we also want to check that unchanged parts are actually unchanged

Why? AFAIU the gaps in the real life may contain any garbade.

Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: @amberhassaan 
> because we're relying on zero being a special value.

Zero is and will forever be a special value in all IEEE-754 compliant FP numbers and their extensions, as well in all two's complement integers. Are we going to support other number formats in the visible future?

File: src/ocl/convolutionocl.cpp
Comment (atamazov) [inline src/ocl/convolutionocl.cpp]: [Recommendation] Let's avoid tautology in the new code.
```suggestion
void ConvolutionDescriptor::ValidateTensors(const ConvTensors& tensors) const
```
BTW maybe it is better to keep this as a freestanding function and pass group_count as a parameter?

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: @amberhassaan ~If you copy reference output buffer into device output buffer before convolution, then verification will pass even if convolution did nothing.~ [irrelevant]

File: src/ocl/convolutionocl.cpp
Comment (averinevg) [inline src/ocl/convolutionocl.cpp:423]: This check must be added to all convolution solvers that do not support non-packed tensors

File: src/include/miopen/convolution.hpp
Comment (averinevg) [inline src/include/miopen/convolution.hpp:39]: [Recommendation] Since this header was added here, it can be removed from src/ocl/convolutionocl.cpp

Or you could use a forward declaration here and not add this header, which is preferable in my opinion. The more includes, the longer the compilation time

File: test/gpu_reference_kernel.cpp
Comment (atamazov) [reply test/gpu_reference_kernel.cpp]: @amberhassaan Aha, now I see. Well, I still think that zeroing is better, but I am Ok with randomizing gaps as well.
[resolved]

**Review by atamazov — COMMENTED** [review]
One non-blocking comment

File: src/ocl/convolutionocl.cpp
Comment (averinevg) [inline src/ocl/convolutionocl.cpp]: Here the functionality overlaps with ValidateGroupCount(). We also have Problem::ValidateGroupCount(). It would be nice to sort this out.

**Review by averinevg — CHANGES_REQUESTED** [review]

File: test/gtest/conv_api_strided_tensors.cpp
Comment (averinevg) [inline test/gtest/conv_api_strided_tensors.cpp:142]: FYI This test was needed to verify that the library returns an error if non-packed tensors are used for convolutions.

Comment (amberhassaan) [reply test/gtest/conv_api_strided_tensors.cpp:142]: Are you OK with this change or should we remove the test altogether?

Comment (averinevg) [reply test/gtest/conv_api_strided_tensors.cpp:142]: This test is no longer needed and can be removed. We need to make a test for non-packed tensors instead.

File: test/gpu_reference_kernel.cpp
Comment (amberhassaan) [reply test/gpu_reference_kernel.cpp:439]: Will remove (next commit once CI completes).

File: src/ocl/convolutionocl.cpp
Comment (amberhassaan) [reply src/ocl/convolutionocl.cpp:423]: @averinevg : What's the best way to go about finding these solvers? How about I update the test to not apply the check if the algorithms are "naive" or "ck" based?

Comment (averinevg) [reply src/ocl/convolutionocl.cpp:423]: @amberhassaan You can find them all in `src/include/miopen/solver.hpp`. It is better to implement this check inside solvers' `IsApplicable()` because this is exactly the place where such checks should be made.

Comment (averinevg) [reply src/ocl/convolutionocl.cpp:423]: A general function for checking can be added to `conv::ProblemDescription`

Comment (amberhassaan) [reply src/ocl/convolutionocl.cpp:423]: Got it. However, due to shortage of time, I'd like an easier solution. It seems like we have many many solvers for convolution and updating all of them will take time.
@JehandadKhan , @junliume : thoughts? context: we need to disable the check above to allow non-packed tensors in convolution, but then the check is needed for all the solvers that don't support non-packed tensors, and those are like 99% of them.

File: test/gtest/conv_api_strided_tensors.cpp
Comment (amberhassaan) [reply test/gtest/conv_api_strided_tensors.cpp:142]: True. We don't have a public API test for non-packed tensors. I have modified reference convolution tests to generate non-packed strides in this PR and #2418 will add tests for 3D convolution, but those target specific solvers. In general, it seems we lack public API tests (or is it my misconception?). A lot of tests seem to use internal APIs.

File: src/ocl/convolutionocl.cpp
Comment (JehandadKhan) [reply src/ocl/convolutionocl.cpp:423]: In `conv::ProblemDescription` create a new helper function which checks if all the tensor ( input. output, weights) are packed. Then in each solver's `IsApplicable` call it to enforce the condition. While I agree it is cumbersome but in the current implementation, that is how all of this is implemented. For example, when a new layout/data type is added, we have go through all solver and make sure the old solvers respect the new feature.

Comment (amberhassaan) [reply src/ocl/convolutionocl.cpp:423]: ok, I agree with all of that. I just want to postpone this change to after our deadline.

**Comment by junliume** [comment]
@CAHEK7 @averinevg @DrizztDoUrden @atamazov @JehandadKhan @iq136boy :
CI has passed, last call for any blocking reviews before merging.

File: src/ocl/convolutionocl.cpp
Comment (amberhassaan) [reply src/ocl/convolutionocl.cpp]: https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2444 created to handle this. 

Proposal: We remove `Problem::ValidateGroupCount()` and then combine `ValidateGroupCount()` in `convolutionocl` with `conv::ProblemDescription::ValidateTensors()` (currently `ConvDescriptor::ValidateTensors()` and use it everywhere. This will allow getting `N` and `C` lengths and strides easily from accessors in `conv::ProblemDescrioption`. 

Alternate Proposal: Combine all `ValidateGroupCount` with `ConvDescriptor::ValidateTensors()`. Downside: this requires logic to extract `N` and `C` lengths and strides.

File: src/include/miopen/convolution.hpp
Comment (amberhassaan) [reply src/include/miopen/convolution.hpp:39]: https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2444 created to handle this in case @junliume wants to merge this PR as is (due to urgency).

File: src/ocl/convolutionocl.cpp
Comment (amberhassaan) [reply src/ocl/convolutionocl.cpp:423]: https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2444 created to handle this as it's a more involved change. @junliume @JehandadKhan

File: src/ocl/convolutionocl.cpp
Comment (amberhassaan) [reply src/ocl/convolutionocl.cpp]: https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2444 created to handle this in case we decide to merge the PR as is: @junliume

**Comment by amberhassaan** [comment]
@junliume : https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2444 created to track the unresolved stuff.

**Review by CAHEK7 — DISMISSED** [review]
Seems fine, but conflicting with #2418 and I'm not sure which one has to be merged first.

**Comment by amberhassaan** [comment]
> Seems fine, but conflicting with #2418 and I'm not sure which one has to be merged first.

#2418 was based on top of #2394 (this PR), so this should merge first.

File: src/ocl/convolutionocl.cpp
Comment (averinevg) [inline src/ocl/convolutionocl.cpp]:
```suggestion
        if(g_stride_index != -1)
```

File: src/ocl/convolutionocl.cpp
Comment (averinevg) [inline src/ocl/convolutionocl.cpp:404]: Let's add exception for other layouts

**Review by averinevg — DISMISSED** [review]

File: test/gtest/conv_api_strided_tensors.cpp
Comment (averinevg) [reply test/gtest/conv_api_strided_tensors.cpp:142]: Let's comment out this test for now and create an issue to handle the correct test for non-packed tensors using the same cpp

File: src/include/miopen/convolution.hpp
Comment (amberhassaan) [reply src/include/miopen/convolution.hpp:39]: Done in latest commit.

File: src/ocl/convolutionocl.cpp
Comment (amberhassaan) [reply src/ocl/convolutionocl.cpp]: Done in latest commit.

**Review by junliume — APPROVED** [review]

**Pull request merged**
